frame/safe-mode/src/lib.rs

```diff
@@ -0,0 +1,439 @@
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Currency type for this pallet.
+		type Currency: ReservableCurrency<Self::AccountId>;
+
+		/// Contains all calls that can be dispatched even when the safe-mode is activated.
+		///
+		/// The `SafeMode` pallet is always included and does not need to be added here.
+		type SafeModeFilter: Contains<Self::RuntimeCall>;
+
+		/// How long the safe-mode will stay active when activated with [`Pallet::activate`].
+		#[pallet::constant]
+		type ActivateDuration: Get<Self::BlockNumber>;
+
+		/// For how many blocks the safe-mode can be extended by each [`Pallet::extend`] call.
+		///
+		/// This does not impose a hard limit as the safe-mode can be extended multiple times.
+		#[pallet::constant]
+		type ExtendDuration: Get<Self::BlockNumber>;
+
+		/// The amount that will be reserved upon calling [`Pallet::activate`].
+		///
+		/// `None` disallows permissionlessly enabling the safe-mode.
+		#[pallet::constant]
+		type ActivateStakeAmount: Get<Option<BalanceOf<Self>>>;
+
+		/// The amount that will be reserved upon calling [`Pallet::extend`].
+		///
+		/// `None` disallows permissionlessly extending the safe-mode.
+		#[pallet::constant]
+		type ExtendStakeAmount: Get<Option<BalanceOf<Self>>>;
+
+		/// The origin that can call [`Pallet::force_activate`].
+		///
+		/// The `Success` value is the number of blocks that this origin can activate safe-mode for.
+		type ForceActivateOrigin: EnsureOrigin<Self::Origin, Success = Self::BlockNumber>;
+
+		/// The origin that can call [`Pallet::force_extend`].
+		///
+		/// The `Success` value is the number of blocks that this origin can extend the safe-mode.
+		type ForceExtendOrigin: EnsureOrigin<Self::Origin, Success = Self::BlockNumber>;
+
+		/// The origin that can call [`Pallet::force_inactivate`].
+		type ForceInactivateOrigin: EnsureOrigin<Self::Origin>;
+
+		/// The origin that can call [`Pallet::repay_stake`] and [`Pallet::slash_stake`].
+		type RepayOrigin: EnsureOrigin<Self::Origin>;
+
+		// Weight information for extrinsics in this pallet.
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The safe-mode is (already or still) active.
+		IsActive,
+
+		/// The safe-mode is (already or still) inactive.
+		IsInactive,
+
+		/// A value that is required for the extrinsic was not configured.
+		NotConfigured,
+
+		/// There is no balance staked.
+		NotStaked,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// The safe-mode was activated until inclusively this \[block\].
+		Activated(T::BlockNumber),
```

Comment:
can we use the named parameter syntax for all the new pallets.
i.e. `Activated { block: T::BlockNumber },`

Reply:
included in #12371 (but it that isn't merged, will bring in here À la carte)